sdk/inc/azure/iot/az_iot_adu_client.h

```diff
@@ -0,0 +1,620 @@
+
+/**
+ * @brief The root key id used to identify the key.
+ */
+extern const uint8_t azure_iot_adu_root_key_id[13];
+
+/**
+ * @brief The root key n (modulus) used to verify the manifest.
+ */
+extern const uint8_t azure_iot_adu_root_key_n[385];
+
+/**
+ * @brief The root key e (exponent) used to verify the manifest.
+ */
+extern const uint8_t azure_iot_adu_root_key_e[3];
+
+/**
+ * @brief     Identity of the update request.
+ * @remark    This version refers to the update request itself.
+ *            For verifying if an update request is applicable to an
+ *            ADU agent, use the update manifest instructions steps "installed criteria".
+ */
+typedef struct
+{
+  /**
+   * The provider for the update.
+   */
+  az_span provider;
+  /**
+   * The name for the update.
+   */
+  az_span name;
+  /**
+   * The version for the update.
+   */
+  az_span version;
+} az_iot_adu_update_id;
+
+/**
+ * @brief Holds any user-defined custom properties of the device.
+ * @remark Implementer can define other device properties to be used
+ *         for the compatibility check while targeting the update deployment.
+ */
+typedef struct
+{
+  /**
+   * An array holding the custom names for the device properties.
+   */
+  az_span names[AZ_IOT_ADU_CLIENT_MAX_DEVICE_CUSTOM_PROPERTIES];
+  /**
+   * An array holding the custom values for the device properties.
+   */
+  az_span values[AZ_IOT_ADU_CLIENT_MAX_DEVICE_CUSTOM_PROPERTIES];
+  /**
+   * The number of custom names and values.
+   */
+  int32_t count;
+} az_iot_adu_device_custom_properties;
+
+/**
+ * @brief      Holds the ADU agent device properties.
+ * @remarks    These properties are used by the ADU service for matching
+ *             update groups and verifying the current update deployed.
+ * https://docs.microsoft.com/azure/iot-hub-device-update/device-update-plug-and-play
+ */
+typedef struct
+{
+  /**
+   * The device manufacturer of the device, reported through deviceProperties.
+   */
+  az_span manufacturer;
+  /**
+   * The device model of the device, reported through deviceProperties.
+   */
+  az_span model;
+  /**
+   * Implementer can define other device properties to be used for the
+   * compatibility check while targeting the update deployment.
+   */
+  az_iot_adu_device_custom_properties* custom_properties;
+  /**
+   * Version of the Device Update agent running on the device.
+   * @remark Must be set to AZ_IOT_ADU_CLIENT_AGENT_VERSION.
+   */
+  az_span adu_version;
+  /**
+   * Version of the Delivery Optimization agent.
+   * @remark Please see Azure Device Update documentation on how to use
+   *         the delivery optimization agent. If unused, set to #AZ_SPAN_EMPTY.
+   *
+   * https://docs.microsoft.com/azure/iot-hub-device-update/device-update-plug-and-play#device-properties
+   */
+  az_span delivery_optimization_agent_version;
+  /**
+   * An ID of the update that is currently installed.
+   */
+  az_span update_id;
+} az_iot_adu_client_device_properties;
```

Comment:
I think this is large enough to warrant initialization APIs. That way, if the app doesn't provide some of the values, the constructor will be able to fill-in defaults.

Reply:
these are usually statically allocated and remain unchanged for the device.

Reply:
https://github.com/Azure/azure-sdk-for-c/pull/2392